### PR TITLE
Add Repository Variables and Secrets

### PR DESCRIPTION
**Pull request by aochsner**
## Hey, I just made a Pull Request!

Adding Repository Variables and Secrets to github scaffolder actions

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: packages/integration/package.json [hunk 1]
```diff
@@ -34,7 +34,7 @@
   "dependencies": {
     "@backstage/config": "workspace:^",
     "@backstage/errors": "workspace:^",
-    "@octokit/auth-app": "^4.0.0",
+    "@octokit/auth-app": "4.0.9",
```
Comment (freben) [inline packages/integration/package.json:37]: This needs a changeset. But can we revert these? Are the changes really necessary? You didn't provide an explanation for what you're achieving.

If this is about the same thing that [these resolutions](https://github.com/backstage/backstage/blob/667c490cc10fdb6dea06ccfe7ef2ab92364d7d45/packages/create-app/templates/default-app/package.json.hbs#L40) address, then you should probably revert these bumps and use resolutions for now. The octokit types are just broken temporarily and we're waiting for them to ship a fix.

File: plugins/scaffolder-backend/package.json [hunk 2]
```diff
@@ -81,12 +81,13 @@
     "isomorphic-git": "^1.23.0",
     "jsonschema": "^1.2.6",
     "knex": "^2.0.0",
+    "libsodium-wrappers": "^0.7.11",
```
Comment (freben) [inline plugins/scaffolder-backend/package.json:84]: Hm this adds almost 200 kB. Is there any way we can use the builtin Crypto module in node or so?

File: plugins/scaffolder-backend/src/scaffolder/actions/builtin/github/helpers.ts [hunk 3]
```diff
@@ -254,6 +258,51 @@ export async function createGithubRepoWithCollaboratorsAndTopics(
     }
   }
 
+  for (const key in environmentVariables) {
+    if ({}.hasOwnProperty.call(environmentVariables, key)) {
```
Comment (freben) [inline plugins/scaffolder-backend/src/scaffolder/actions/builtin/github/helpers.ts:262]:
```suggestion
    if (Object.hasOwnProperty.call(environmentVariables, key)) {
```

hehe, no need to change this but constructing a new object each time just feels wrong :)

Would also feel even better with `for (const [key, value] of Object.entries(environmentVariables)) {...`

**Review by freben — COMMENTED** [review]
Thanks!

File: plugins/scaffolder-backend/package.json (hunk 2 above)
Comment (aochsner) [reply plugins/scaffolder-backend/package.json:84]: Potentially?  Just following GitHub doco directly:
- https://docs.github.com/en/rest/actions/secrets?apiVersion=2022-11-28#create-or-update-a-repository-secret
- https://docs.github.com/en/actions/security-guides/encrypted-secrets

Thoughts?

Comment (freben) [reply plugins/scaffolder-backend/package.json:84]: Ah I think that makes a decent argument for keeping it as is.

File: package.json [hunk 4]
```diff
@@ -45,7 +45,8 @@
   },
   "resolutions": {
     "@types/react": "^17",
-    "@types/react-dom": "^17"
+    "@types/react-dom": "^17",
+    "@octokit/auth-app": "4.0.9"
```
Comment (freben) [inline package.json:49]: remove

File: packages/create-app/templates/default-app/package.json.hbs [hunk 5]
```diff
@@ -42,7 +42,8 @@
     "@types/react-dom": "^17",
     "@octokit/rest": "18.5.6",
     "@octokit/plugin-rest-endpoint-methods": "5.16.2",
-    "@octokit/plugin-paginate-rest": "2.7.0"
+    "@octokit/plugin-paginate-rest": "2.7.0",
+    "@octokit/auth-app": "4.0.9"
```
Comment (freben) [inline packages/create-app/templates/default-app/package.json.hbs:46]: remove

File: .changeset/wet-vans-drive.md [hunk 6]
```diff
@@ -0,0 +1,5 @@
+---
+'@backstage/plugin-scaffolder-backend': minor
+---
+
+Add Repository Variables and Secrets to scaffolder action. Bump octokit version.
```
Comment (freben) [inline .changeset/wet-vans-drive.md:5]:
```suggestion
Add support for Repository Variables and Secrets to the `publish:github` and `github:repo:create` scaffolder actions.
```

**Review by freben — COMMENTED** [review]
Just minor cleanup, then good to go.

**Comment by freben** [comment]
Closes https://github.com/backstage/backstage/pull/17722

**Comment by aochsner** [comment]
Ok I need some help w/ these errors.  Not sure the right node/typescript way to solve this.

**Comment by freben** [comment]
Ugh. Still being affected by https://github.com/octokit/rest.js/issues/290 it seems. Is this rebased on the very latest master?

**Comment by aochsner** [comment]
> Ugh. Still being affected by [octokit/rest.js#290](https://github.com/octokit/rest.js/issues/290) it seems. Is this rebased on the very latest master?

Nope.  Just rebased.  Lets see what transpires

**Review by freben — COMMENTED** [review]
OK I think this is ready to go. As you'll see, there's a conflict with yarn.lock now. I'm noting that you keep merging master into your branch, and you may be able to make that work too, but it might be easier to squash all of this into one commit instead, and resetting yarn.lock to the newest master version and then running `yarn install` again to let yarn resolve it all magically for you instead of trying to merge-fix which tends to become frustrating and hard to track.

**Comment by aochsner** [comment]
Did as you suggested.  Still having type compilation issues.  I need at least `2.0.3` of `octokit` to get `createRepoVariable()` but every time I touch that it seems to spew all sorts of tsc issues.  I can't seem to find the magic combination

**Comment by Pike** [comment]
Nice, I was just looking for this.

Does this change the required scope on the GH tokens and apps?

**Comment by Pike** [comment]
I'd say for one in the changelog, so that it's part of the release notes.

Also, https://backstage.io/docs/integrations/github/locations#token-scopes and https://backstage.io/docs/integrations/github/github-apps#app-permissions.

Maybe @ahhhndre has some more tips, I know he struggled with the last change that flew under the radar.

**Review by freben — APPROVED** [review]

**Pull request merged**